### ISSUE TITLE
Analyzer: Improperly failed inference with promoted type parameters

### ISSUE THREAD

**Issue by MichaelRFairhurst** [dart-lang/sdk#35799]
Here, the call to f should be valid and inferrable as `Cloneable<T>`.

```
class Clonable<T> {}                                                                             
class SubClonable<T> extends Clonable<T> {
  T m(T t) => t;  
} 

void f<T extends Clonable<T>>(T object) {
  if (object is SubClonable<T>) {
    f(object);
  }     
}
```

However it fails in analyzer and CFE. This is the analyzer ticket for this issue. (CFE is #35800)

It's critical to note that `object` is promoted to type `T' extends SubCloneable<T>`. So if the analyzer attempts to infer `SubClonable<T>` for `T` in `f(object)`, then that will fail. However, it should be possible to infer Clonable<T>, which in fact works when typed explicitly.

May be related to #35100

**Comment by stereotype441** [comment]
Note: there is a test in the analyzer codebase that should be re-enabled when this bug is fixed: https://dart-review.googlesource.com/c/sdk/+/90880/4/pkg/analyzer/test/src/task/strong/checker_test.dart#3989

**Comment by eernstg** [comment]
I'm not convinced that this issue should be considered as a bug in the analyzer, it is probably about the type inference algorithm. I created https://github.com/dart-lang/language/issues/3009 for that.

**Comment by FMorschel** [comment]
So, just to help out a bit, as for what I found: 

#52077
https://github.com/dart-lang/linter/issues/4277
https://github.com/dart-lang/language/issues/3012 (this is potentially the solution for the other two)

Are all possibly related?

**Comment by eernstg** [comment]
@FMorschel, thanks for looking into those potential connections!

I do think there is a connection between dart-lang/linter#4277 and #52077 and dart-lang/language#3012, they are all about extensions and intersection types.

However, there is probably no connection to this issue, where the topic is F-bounded type variables and inference (with no references to intersection types or extensions at all).

**Comment by gnprice** [comment]
The language issue https://github.com/dart-lang/language/issues/3009 (filed for this as mentioned above) has been resolved. Does this issue still apply?

**Comment by eernstg** [comment]
Good catch! The issue has indeed been resolved in https://dart-review.googlesource.com/c/sdk/+/364721.
